### ISSUE TITLE
segfault on exiting R after re-running model with multiple cores

### ISSUE THREAD

**Issue by Rharald** [stan-dev/rstan#296]
#### Summary:

When exiting R after re-running a model there's a segfault when 
options(mc.cores = parallel::detectCores()) is set. No segfault when either running the same code with a single core or re-running in a new session.
#### Description:

When exiting R after re-running a model there's a segfault when 
options(mc.cores = parallel::detectCores()) is set. No segfault when either running the same code with a single core or re-running in a new session.
#### Reproducible Steps:

library(rstan)
rstan_options(auto_write = TRUE)
options(mc.cores = parallel::detectCores())

run example 1 from ?stan

scode <- "
  parameters {
  real y[2];
  }
  model {
    y[1] ~ normal(0, 1);
    y[2] ~ double_exponential(0, 2);
  }
"
fit1 <- stan(model_code = scode, iter = 10, verbose = FALSE)
fit2 <- stan(fit = fit1, iter = 10000, verbose = FALSE)

exit R
Save workspace image? [y/n/c]: n

 **\* caught segfault ***
address 0x7f152d78da40, cause 'memory not mapped'

Traceback:
 1: .Call(list(name = "CppObject__finalize", address = <pointer: 0x19c9e30>,     dll = l$
 2: x$.self$finalize()
 3: (function (x) x$.self$finalize())(<environment>)

Possible actions:
1: abort (with core dump, if enabled)
2: normal R exit
3: exit R without saving workspace
4: exit R saving workspace

When I leave the session after "fit1 <- ..." with saving the workspace the "fit2 <-" runs without error in a restored session, and I can quit without segfaulting.

When I run the same code without the 'parallel' option everything's okay (no segfault on exit).

> sessionInfo()
> R version 3.3.0 (2016-05-03)
> Platform: x86_64-pc-linux-gnu (64-bit)
> Running under: Debian GNU/Linux stretch/sid

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C  
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8  
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8  
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C  
 [9] LC_ADDRESS=C               LC_TELEPHONE=C  
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] rstan_2.9.0-3 ggplot2_2.1.0

loaded via a namespace (and not attached):
 [1] colorspace_1.2-6 scales_0.4.0     plyr_1.8.3       tools_3.3.0  
 [5] parallel_3.3.0   inline_0.3.14    gtable_0.2.0     gridExtra_2.2.1 
 [9] Rcpp_0.12.4      codetools_0.2-14 grid_3.3.0       stats4_3.3.0  
[13] munsell_0.4.3   
#### additional information

I ran the same code using Dirk Eddelbuettel's littler ("r") and get this in /var/log/syslog

"[...] kernel: [54200.845744] r[20628]: segfault at 7f19bb5f5a40 ip 00007f19be31ff79 sp 0007fff15a23490 error 4 in Rcpp.so[7f19be303000+53000]"

So maybe there's an incompatibility between the CRAN versions of Rcpp and rstan using the latest R version.

**Comment by bgoodri** [comment]
I saw this too a few weeks ago. Then it went away. Now it seems to be back. I think it is more likely to be a Debian or a Rcpp thing than a rstan thing, but I'll leave this open until someone figures something out.

**Comment by bgoodri** [comment]
Upon further review, I think it is a R change for 3.3.0 that adversely affects rstan. @maverickg Can you look at it? To replicate, I think it is sufficient to just reuse a stanmodel that has been called in parallel, as in

```
library(rstan)
funnel <- stan_demo("funnel", cores = 2)
funnel2 <- stan(fit = funnel, cores = 1)
q("no") # segfault
```

**Comment by bgoodri** [comment]
@maverickg I have avoided the segfault on Linux by changing 
`reg.finalizer(environment(), cleanup, onexit = TRUE)`
to
`reg.finalizer(environment(), cleanup, onexit = FALSE)`
in the cxxfun_from_dso_bin() function in the cxxfunplus.R file.
Do you remember a reason for `onexit` to be `TRUE`?

**Comment by dhaine** [comment]
Any update on this issue? I got the same segfault after upgrading to R 3.3.0 on Amazon AWS (Ubuntu 14.04). Or is it a different issue?

 **\* caught segfault ***
address 0x7fa294a31648, cause 'memory not mapped'

Traceback:
 1: .Call(list(name = "CppObject__finalize", address = <pointer: 0x2d44d40>,     dll = list(name = "Rcpp", path = "/usr/lib/R/site-library/Rcpp/libs/Rcpp.so",         dynamicLookup = TRUE, handle = <pointer: 0x25e3d90>,         info = <pointer: 0x7fa2eb8b8e80>), numParameters = 2L),     <pointer: 0x7e15120>, .pointer)
 2: x$.self$finalize()
 3: (function (x) x$.self$finalize())(<environment>)
An irrecoverable exception occurred. R is aborting now ...
Segmentation fault (core dumped)

**Comment by bgoodri** [comment]
It was ostensibly fixed by b6307b133539b983bac561185e5b5b988c69c2d3

**Comment by maverickg** [comment]
The reason it was TRUE in the original code is that I mimicked from https://github.com/cran/inline/blob/master/R/cxxfunction.R#L145

**Comment by bgoodri** [comment]
This should be fixed in 2.10.0, which is now on CRAN in source form at least.

**Comment by Rharald** [comment]
I can confirm the issue is fixed in Version 2.10.1 which I just pulled from CRAN.

Thanks.
